### DISCUSSION TITLE
best structure for ci-cd

### DISCUSSION THREAD

**Discussion by quantum333** [beeware/toga#2303] [Q&A]
Hi

i would really appreciate if you could give me your opinion on some of the questions/issues below.  

Good karma awaiting for you!

I am the owner of small company with fair but no high end tech knowledge in need of some advice on few issues of structure for future developments in terms of CI-cd in this set up.

currently i have a one part time developer on maintenance "mission" for: 

1. a website on a main domain 
2. an crm for internal use built on a subdomain. This crm has been developed from scratch over a period of 10 years. It can update partially the website for news, faqs, etc and for the internal agents (sales)  has an inbox, contact & appointment management  and some very specific tools of our business connected to an APIs. This  crm serves as a tool to the internal agent (sales) to provide information and close deals by phone or email. 

It is all hosted on a server self-managed using 70 GB all included. 

We want to create a structure to the developments ahead so we end up with a more efficient way to do business, received and share data with clients, set up alerts...  So we wish to build a 

4. virtual office for clients. So the crm (internal agents) can push information and interact with the clients with things such as communication, price offers, documents of interest, etc. 
5. mobile app which is pretty much the virtual office number 3 but on a mobile app

The advice provided so far is:

Step 1. Create 4 different repositories in github and as many branches as needed. 
Step 2. On a new server open 4 different test sites or subdomain to test each of the 4 jobs at hand. 

Not sure this is the best environment to the challenges ahead. My concerns so far

A. own server v.s serverless. ( I like knowing what i pay monthly, and i know it is difficult to calculate what you pay for a serverless at the end of the month) ( also, i  understand that once you go into serverless it is more difficult to go back to own server if you wish so) 

B. Development. ( developers often show me codes but i can not visually see what they will look on a web or the crm until they go to the test site which sometimes takes time.  Is there a tool for me to see the visual form of that code instead of waiting to be pushed to the test site?) 

C. virtual office/mobile application. Basically they will server the same purpose. Can we have androip app / apple app / virtual office on the web as a one package on the development?) 

D. Testing. open a new server and have different repositories which are all interconected in the sense that the crm pushes data to the website and in the future to the clients virtual office and the application.. not sure it sounds like a good idea. 

E. Monitoring and feedback. i can foresee also issues unless we have an integrated set up to record and follow feedback of bugs, etc in terms of test enviroment but also in production. (agents currently report issues via slack but are not registered properly since they are all answer by the same current developer) Any tips on how to manage this feedback without been a real hassle?)

Thanks in advance for your comments

**Comment by freakboy3742** [comment]
It's not at all clear why you're asking this question on the Toga discussion forum - or, for that matter... what you're asking for advice on.

Toga is a GUI toolkit. It's not clear how any of your questions about a "virtual office" (whatever that is) relate to Toga. Questions about web server deployment are definitely off topic... Toga *has* a web backend, but a Toga web deployment is essentially a static page.

There's a couple of side references to a mobile app... are you intending to build an app with Toga as part of this "virtual office" configuration? If so, we might be able to provide some advice about that... but if this is what you're looking for, it's not at all clear what your question actually *is*.

And, the overall question is framed in terms of CI/CD setup... but then none of the problems that you describe are CI/CD related (as far as I can make out).

It seems like you've thrown a question to a random forum in the hope of getting a response. If I've misunderstood what you're asking, then let me know; otherwise, I suspect you may need to take your question to a forum that is more closely aligned with your requirements. Something associated with your web framework or web server of choice might be a closer match.
